src/enclave/enclave_oe.c

```diff
@@ -48,6 +55,54 @@ static int _strncmp(const char* x, const char* y, size_t n)
     return *px == *py ? 0 : *px < *py ? -1 : +1;
 }
 
+static void
+init_auxv(size_t *auxv, char* buf_ptr, char *pn)
+{
+    // By default auxv[AT_RANDOM] points to a buffer with 16 random bytes.
+    uint64_t *rbuf = (uint64_t*)buf_ptr;
+    buf_ptr += 16;
+    // TODO Use intrinsics
+    // if (!_rdrand64_step(&rbuf[0]))
+    //    goto err;
+    register uint64_t rd;
+    __asm__ volatile("rdrand %0;"
+                     : "=r"(rd));
+    rbuf[0] = rd;
+    __asm__ volatile("rdrand %0;"
+                     : "=r"(rd));
+    rbuf[1] = rd;
+
+    memset(auxv, 0, 2 * sizeof(size_t) * AUXV_ENTRIES);
+    auxv[0] = AT_CLKTCK;
+    auxv[1] = 100;
+    auxv[2] = AT_EXECFN;
+    auxv[3] = (size_t) pn;
+    auxv[4] = AT_HWCAP;
+    auxv[5] = 0;
+    auxv[6] = AT_EGID;
+    auxv[7] = 0;
+    auxv[8] = AT_EUID;
+    auxv[9] = 0;
+    auxv[10] = AT_GID;
+    auxv[11] = 0;
+    auxv[12] = AT_PAGESZ;
+    auxv[13] = 0;
+    auxv[14] = AT_PLATFORM;
+    memcpy(buf_ptr, at_platform, oe_strlen(at_platform) + 1);
+    auxv[15] = (size_t) buf_ptr;
+    buf_ptr += oe_strlen(at_platform) + 1;
+    auxv[16] = AT_SECURE;
+    auxv[17] = 0;
+    auxv[18] = AT_UID;
+    auxv[19] = 0;
+    auxv[20] = AT_RANDOM;
+    auxv[21] = (size_t)rbuf;
+    auxv[22] = AT_HW_MODE;
+    auxv[23] = !sgxlkl_in_sw_debug_mode();
+    auxv[24] = AT_NULL;
+    auxv[25] = 0;
+}
```

Comment:
The following seem to be missing: 
- AT_BASE -- base address of image
- AT_SYSINFO_EHDR -- the ELF header address
- AT_PHDR - the program header address
- AT_PHNUM - the number of program headers
 - AT_ENTRY - address of the entry point

These are usually passed by the loader. It may work without them for now, but these may be needed later when we want to just call _start().

You can check in without these I think.

Reply:
`AT_PHDR` is needed for unwind libraries.  We probably can't throw an exception through `qsort` without that.  That said, I've never actually seen code in the wild that wants to throw exceptions through libc functions outside of test suites: if you're using C++, there are significantly more efficient ways of sorting than calling libc's `qsort` and very few other libc functions take function pointer arguments.